File: backend/src/main/java/pers/eastwind/billmanager/user/controller/UserController.java
package pers.eastwind.billmanager.user.controller;

import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import pers.eastwind.billmanager.common.model.PageResult;
import pers.eastwind.billmanager.common.model.QueryParam;
import pers.eastwind.billmanager.user.config.UserConfigProperties;
import pers.eastwind.billmanager.user.model.LoginResult;
import pers.eastwind.billmanager.user.model.UserDTO;
import pers.eastwind.billmanager.user.service.UserService;

import java.util.List;

/**
 * 用户控制器
 */
@RestController
@RequestMapping("/api/user")
public class UserController {
    private final UserConfigProperties config;
    private final UserService userService;

    public UserController(UserConfigProperties config, UserService userService) {
        this.config = config;
        this.userService = userService;
    }

    /**
     * 登录
     * 设置 token 至响应头
     *
     * @param param 登录参数
     * @return 用户信息
     */
    @PostMapping("/token")
    public ResponseEntity<UserDTO> login(@RequestBody LoginParam param) {
        LoginResult res = userService.login(param.username, param.password);
        return ResponseEntity.ok().header("Set-Cookie",
                        ResponseCookie.from("X-Auth-Token", res.token())
                                .path("/")
                                .httpOnly(true)
                                .secure(true)
                                .maxAge(config.getExpire())
                                .sameSite("Strict").build().toString())
                .body(res.user());
    }

    /**
     * 查询
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'USER', 'FINANCE')")
    @GetMapping
    public PageResult<UserDTO> getAll(QueryParam queryParam) {
        return userService.getAll(queryParam);
    }

    /**
     * 创建
     */
    @PreAuthorize("hasAnyRole('ADMIN')")
    @PostMapping
    public UserDTO create(@RequestBody UserDTO user) {
        return userService.create(user);
    }

    /**
     * 修改
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'USER', 'FINANCE')")
    @PutMapping
    public UserDTO update(@RequestBody UserDTO user) {
        return userService.update(user);
    }

    /**
     * 禁用
     * @param username 用户名
     */
    @DeleteMapping("/{username}")
    @PreAuthorize("hasAnyRole('ADMIN')")
    public void disable(@PathVariable String username) {
        userService.disable(username);
    }

    /**
     * 登录参数
     *
     * @param username 用户名
     * @param password 密码
     */
    public record LoginParam(String username, String password) {
    }
}

File: backend-go/internal/user/controller/user_controller.go
package controller

import (
	"backend-go/config"
	"backend-go/internal/common/errs"
	"backend-go/internal/common/result"
	"backend-go/internal/user/model"
	"backend-go/internal/user/service"

	"github.com/gofiber/fiber/v2"
)

// UserController 用户控制器
type UserController struct {
	cfg         *config.JWTConfig
	userService *service.UserService
}

// NewUserController 创建用户控制器实例
func NewUserController(cfg *config.JWTConfig, userService *service.UserService) *UserController {
	return &UserController{
		userService: userService,
		cfg:         cfg,
	}
}

// Login 用户登录
func (c *UserController) Login(ctx *fiber.Ctx) error {
	var param model.LoginParam
	err := ctx.BodyParser(&param)
	if err != nil {
		return err
	}
	origin := ctx.Get("Origin")
	res, err := c.userService.Login(ctx.Context(), param.Username, param.Password, origin)
	if err != nil {
		return err
	}
	cookie := fiber.Cookie{
		Name:     "X-Auth-Token",
		Value:    res.Token,
		Path:     "/",
		MaxAge:   c.cfg.Expire,
		Secure:   true,
		SameSite: "Strict",
		HTTPOnly: true,
	}
	ctx.Cookie(&cookie)
	result.SetResult(ctx, res)
	return nil
}

// GetAll 获取用户列表
func (c *UserController) GetAll(ctx *fiber.Ctx) error {
	var param result.QueryParam
	err := ctx.QueryParser(&param)
	if err != nil {
		return err
	}
	res, err := c.userService.GetAll(ctx.Context(), &param)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

// Create 创建用户
func (c *UserController) Create(ctx *fiber.Ctx) error {
	var dto model.UserDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}

	res, err := c.userService.Create(ctx.Context(), &dto)
	if err != nil {
		return err
	}

	result.SetResult(ctx, res)
	return nil
}

// Update 更新用户
func (c *UserController) Update(ctx *fiber.Ctx) error {
	var dto model.UserDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}
	res, err := c.userService.Update(ctx.Context(), &dto)
	if err != nil {
		return err
	}

	result.SetResult(ctx, res)
	return nil
}

// Disable 禁用用户
func (c *UserController) Disable(ctx *fiber.Ctx) error {
	username := ctx.Params("username")
	if username == "" {
		return errs.NewBizError("用户名为空")
	}

	if err := c.userService.Disable(ctx.Context(), username); err != nil {
		return err
	}

	return nil
}
